executor/pipeline/fetch.rs

```diff
@@ -316,27 +404,74 @@ fn execute_attributes_list<'a>(
-                        attribute.as_reference().into_owned(),
-                    )))));
-                }
+        let (has, count) = result.map_err(|source| FetchExecutionError::ConceptRead { source })?;
+        let attribute = has.attribute();
+        let suitable = attribute
+            .type_()
+            .is_subtype_transitive_of_or_same(snapshot.as_ref(), thing_manager.type_manager(), attribute_type.clone())
+            .map_err(|source| FetchExecutionError::ConceptRead { source })?;
+        if suitable {
+            for _ in 0..count {
+                list.list.push(DocumentNode::Leaf(DocumentLeaf::Concept(Concept::Thing(Thing::Attribute(
+                    has.attribute().into_owned(),
+                )))));
             }
-            Err(err) => return Err(FetchExecutionError::ConceptReadError { source: err }),
         }
     }
     Ok(list)
 }
 
+fn prepare_attribute_type_has_iterator<'a>(
+    object: impl ObjectAPI<'a>,
+    attribute_type: AttributeType<'static>,
+    snapshot: &Arc<impl ReadableSnapshot>,
+    thing_manager: &Arc<ThingManager>,
+) -> Result<HasIterator, FetchExecutionError> {
+    let subtypes = attribute_type
+        .get_subtypes_transitive(snapshot.as_ref(), thing_manager.type_manager())
+        .map_err(|source| FetchExecutionError::ConceptRead { source })?;
+    let attribute_types = TypeAPI::chain_types(attribute_type.clone(), subtypes.into_iter().cloned());
+
+    object
+        .get_has_types_range_unordered(snapshot.as_ref(), thing_manager.as_ref(), attribute_types)
+        .map_err(|source| FetchExecutionError::ConceptRead { source })
+}
+
+fn execute_single_function<Snapshot: ReadableSnapshot + 'static>(
+    snapshot: Arc<Snapshot>,
+    thing_manager: Arc<ThingManager>,
+    parameters: Arc<ParameterRegistry>,
+    functions_registry: Arc<ExecutableFunctionRegistry>,
+    variable_positions: &HashMap<Variable, VariablePosition>,
+    row: MaybeOwnedRow<'_>,
+    function: &ExecutableFunction,
+) -> Result<(PatternExecutor, Arc<ExecutionContext<Snapshot>>), FetchExecutionError> {
+    let mut args = vec![VariableValue::Empty; function.argument_positions.len()];
+    for (var, write_pos) in &function.argument_positions {
+        debug_assert!(write_pos.as_usize() < args.len());
+        args[write_pos.as_usize()] = row.get(variable_positions.get(var).unwrap().clone()).clone().into_owned();
+    }
+    let args = MaybeOwnedRow::new_owned(args, row.multiplicity());
+
+    let step_executors = create_executors_for_function(&snapshot, &thing_manager, &functions_registry, function)
+        .map_err(|err| FetchExecutionError::ConceptRead { source: err })?;
+    let mut pattern_executor = PatternExecutor::new(step_executors);
+    pattern_executor.prepare(FixedBatch::from(args));
+
+    let execution_context =
+        Arc::new(ExecutionContext::new(snapshot.clone(), thing_manager.clone(), parameters.clone()));
```

Comment:
Needless `.clone()`s.
On the other hand, can this function accept an `Arc<ExecutionContext>` rather than reconstruct it?

Reply:
Fooof. I'll try.

Reply:
It sounds massive, so I just removed `clone()`s, thanks!